browser/static/js/adjust.js

```diff
@@ -1,277 +1,398 @@
-      d[i + 1] = d[i+1]*contrast + brightness;
-      d[i + 2] = d[i+2]*contrast + brightness;
+    this.rawLoaded = false;
+    this.segLoaded = false;
+  }
+
+  // getters for brightness/contrast allowed ranges
+  // no setters; these should remain fixed
+  get minBrightness() {
+    return this._minBrightness;
+  }
+
+  get maxBrightness() {
+    return this._maxBrightness;
+  }
+
+  get minContrast() {
+    return this._minContrast;
+  }
+
+  get maxContrast() {
+    return this._maxContrast;
+  }
+
+  // TODO: getters/setters for brightness and contrast?
+  // or use this group of methods?
+
+  changeContrast(inputChange) {
+    const modContrast = -Math.sign(inputChange) * 4;
+    // stop if fully desaturated
+    let newContrast = Math.max(this.contrast + modContrast, this.minContrast);
+    // stop at 8x contrast
+    newContrast = Math.min(newContrast, this.maxContrast);
+
+    if (newContrast !== this.contrast) {
+      // need to retrigger downstream image adjustments
+      this.rawLoaded = false;
+      this.contrast = newContrast;
+      this.contrastRaw();
+    }
+  }
+
+  changeBrightness(inputChange) {
+    const modBrightness = -Math.sign(inputChange);
+    // limit how dim image can go
+    let newBrightness = Math.max(this.brightness + modBrightness, this.minBrightness);
+    // limit how bright image can go
+    newBrightness = Math.min(newBrightness, this.maxBrightness);
+
+    if (newBrightness !== this.brightness) {
+      this.rawLoaded = false;
+      this.brightness = newBrightness;
+      this.contrastRaw();
+    }
+  }
+
+  resetBrightnessContrast() {
+    this.brightness = 0;
+    this.contrast = 0;
+    this.rawLoaded = false;
+    this.contrastRaw();
+  }
+
+  toggleInvert() {
+    this.displayInvert = !this.displayInvert;
+    this.preCompRawAdjust();
+  }
+
+  // modify image data in place to recolor
+  _recolorScaled(data, i, j, jlen, r = 255, g = 255, b = 255) {
+    // location in 1D array based on i and j
+    const pixelNum = (jlen * j + i) * 4;
+    // set to color by changing RGB values
+    // data is clamped 8bit type, so +255 sets to 255, -255 sets to 0
+    data[pixelNum] += r;
+    data[pixelNum + 1] += g;
+    data[pixelNum + 2] += b;
   }
-  return img;
-}
 
-function grayscale(img) {
-  let data = img.data;
-  for (var i = 0; i < data.length; i += 4) {
-      var avg = (data[i] + data[i + 1] + data[i + 2]) / 3;
-      data[i]     = avg; // red
+  // image adjustment functions: take img as input and manipulate data attribute
+  // pixel data is 1D array of 8bit RGBA values
+  // TODO: do we want to pass in B&C values or use object attributes?
+  _contrastImage(img, contrast = 0, brightness = 0) {
+    const d = img.data;
+    contrast = (contrast / 100) + 1;
+    for (let i = 0; i < d.length; i += 4) {
+      d[i] = d[i] * contrast + brightness;
+      d[i + 1] = d[i + 1] * contrast + brightness;
+      d[i + 2] = d[i + 2] * contrast + brightness;
+    }
+  }
```

Comment:
I think we can leave it as taking inputs instead of attributes
```suggestion
  _contrastImage(img, contrast = 0, brightness = 0) {
    const d = img.data;
    contrast = (contrast / 100) + 1;
    for (let i = 0; i < d.length; i += 4) {
      d[i] = d[i] * contrast + brightness;
      d[i + 1] = d[i + 1] * contrast + brightness;
      d[i + 2] = d[i + 2] * contrast + brightness;
    }
  }
```